packages/datadog-instrumentations/src/router.js

```diff
@@ -169,11 +169,100 @@ function createWrapRouterMethod (name) {
 
 const wrapRouterMethod = createWrapRouterMethod('router')
 
-addHook({ name: 'router', versions: ['>=1'] }, Router => {
+addHook({ name: 'router', versions: ['>=1 <2'] }, Router => {
   shimmer.wrap(Router.prototype, 'use', wrapRouterMethod)
   shimmer.wrap(Router.prototype, 'route', wrapRouterMethod)
 
   return Router
 })
 
+addHook({ name: 'router', versions: ['>=2'] }, Router => {
+  const originalRouter = Router
+
+  function WrappedRouter (options) {
+    const router = originalRouter.call(this, options)
+
+    if (!router._queryParsingWrapped) {
+      router._queryParsingWrapped = true
+      const originalHandle = router.handle
+      router.handle = function (req, res, next) {
+        req.query
+        // If query parsing was aborted, don't continue request handling
+        if (res.writableEnded) {
+          return
+        }
+
+        return originalHandle.apply(this, arguments)
+      }
+    }
+
+    return router
+  }
+
+  WrappedRouter.prototype = originalRouter.prototype
+
+  shimmer.wrap(WrappedRouter.prototype, 'use', wrapRouterMethod)
+  shimmer.wrap(WrappedRouter.prototype, 'route', wrapRouterMethod)
+
+  return WrappedRouter
+})
+
+const routerParamStartCh = channel('datadog:router:param:start')
+
+function wrapHandleRequest (original) {
+  return function wrappedHandleRequest (req, res, next) {
+    if (routerParamStartCh.hasSubscribers) {
+      const abortController = new AbortController()
+
+      routerParamStartCh.publish({
+        req,
+        res,
+        params: req?.params,
+        abortController
+      })
+
+      if (abortController.signal.aborted) return
+    }
+
+    return original.apply(this, arguments)
+  }
+}
+
+addHook({
+  name: 'router', file: 'lib/layer.js', versions: ['>=2']
+}, Layer => {
+  shimmer.wrap(Layer.prototype, 'handleRequest', wrapHandleRequest)
+  return Layer
+})
+
+function wrapParam (original) {
```

Comment:
We've tried together and apparently only one i needed :+1: make sure to double check there is no edge case

Reply:
Yes, we need both of them. When using `app.param(param, fn)`, some tests are not working properly:
1. https://github.com/DataDog/dd-trace-js/blob/488fcbfdc495c050f967ade046b8cb592376c1d7/packages/dd-trace/test/appsec/index.express.plugin.spec.js#L168
2. https://github.com/DataDog/dd-trace-js/blob/c44d5392f3f1ee5aa8465f9d78c0de66125626ea/packages/dd-trace/test/appsec/iast/taint-tracking/sources/taint-tracking.express.plugin.spec.js#L185
The issue is because the callback function `fn` is executed inside processParams before reaching `layer.handleRequest`

Reply:
We checked this together, there is no way to have a single instrumentation to process path parameters. However, we introduced a `WeakSet` to avoid calling the path parameters channel twice.